README.md
# Rfm69

Elixir library for configuration and communication with the RFM69 chip

Documentation can be found at [https://hexdocs.pm/rfm69](https://hexdocs.pm/rfm69).

## Installation

```elixir
def deps do
  [
    {:rfm69, "~> 0.1.1"}
  ]
end
```




CHANGELOG.md
## v0.2.0

This release introduces several features. There is now a struct for the Device that holds the relevant information for starting the SPI link. start_link should now be called with the struct. start_link will alias the process with the atom given in the name key of the struct, so it's possible to start more than one chip process at one time now.

When calling any of the public functions on the RFM69 module, the first parameter is the Device struct that was used to call start_link. This allows the RFM69 module to use the name key of the struct to communicate with the process, and also allows upstream libraries to more easily define a protocol for the chip functions using the structs.

Additionally, a new chip_present? function has been added that will attempt to detect whether the chip is responding using the given struct's pin and device configuration.

In some ways the library has gotten more complex because of these changes, but it solves several longstanding problems with supporting different kinds of serial devices and chips, and this should pave the way to making some simplifications in the design for developers using this package as the simplifications become clearer.

## v0.1.2

Updates `elixir_ale` dependency back to fhunleth's hex package with fix for unsetting GPIO interrupts


## v0.1.1

Initial release

This isn't a release complete with support for all the ways to interact with the RFM69 chip or even with full timeout
support or error handling. But it works. Gotta start somewhere!


test/rfm69/configuration_test.exs
defmodule RFM69.ConfigurationTest do
  use ExUnit.Case, async: true

  alias RFM69.Configuration

  test "to_binary returns correct information" do
    # This is a correctness test, which ought to be able to go away once the
    # other behavior tests are in place. For now, it holds the configuration
    # constant to a known good register configuration

    test_configuration = %Configuration{
      frf: Configuration.frequency_to_registers(916_600_000),
      bitrate: Configuration.bitrate_to_registers(16_384),
      data_modul: 0x08,
      pa_level: 0x5F,
      lna: 0x88,
      # 250kHz with dcc freq shift 2, RxBwMant of 16 and RxBwExp of 0
      rx_bw: 0x40,
      # dcc freq shift of 4
      afc_bw: 0x80,
      dio_mapping1: 0x80,
      dio_mapping2: 0x07,
      rssi_thresh: 0xE4,
      preamble: 0x0018,
      sync_config: 0x98,
      sync_value: 0xFF00FF0001010101,
      packet_config1: 0x00,
      payload_length: 0x00,
      fifo_thresh: 0x94,
      packet_config2: 0x00
    }

    {:ok, expected_bytes} =
      Base.decode16(
        "040807A10052E5266641000292F520245F091A40B07B9B884080408006100000000002FF80078000E40000001898FF00FF000101010100000000009400000000000000000000000000000000000100"
      )

    assert expected_bytes == Configuration.to_binary(test_configuration)
  end
end


test/rfm69/rf_configuration_test.exs
defmodule RFM69.RFConfigurationTest do
  use ExUnit.Case, async: true

  alias RFM69.RFConfiguration

  test "to_binary returns correct information" do
    # This is a correctness test, which ought to be able to go away once the
    # other behavior tests are in place. For now, it holds the configuration
    # constant to a known good register configuration

    test_configuration = %RFConfiguration{
      frf: RFConfiguration.frequency_to_registers(916_600_000),
      bitrate: RFConfiguration.bitrate_to_registers(16_384),
      data_modul: 0x08,
      pa_level: 0x5F,
      lna: 0x88,
      rx_bw: 0x40, # 250kHz with dcc freq shift 2, RxBwMant of 16 and RxBwExp of 0
      afc_bw: 0x80, # dcc freq shift of 4
      dio_mapping1: 0x80,
      dio_mapping2: 0x07,
      rssi_thresh: 0xE4,
      preamble: 0x0018,
      sync_config: 0x98,
      sync_value: 0xFF00FF0001010101,
      packet_config1: 0x00,
      payload_length: 0x00,
      fifo_thresh: 0x94,
      packet_config2: 0x00
    }

    {:ok, expected_bytes} = Base.decode16("00040807A10052E5266641000292F520245F091A40B07B9B884080408006100000000002FF80078000E40000001898FF00FF000101010100000000009400000000000000000000000000000000000100")
    assert expected_bytes == RFConfiguration.to_binary(test_configuration)
  end
end


mix.exs
defmodule Rfm69.Mixfile do
  use Mix.Project

  def project do
    [
      app: :rfm69,
      version: "0.2.0",
      elixir: "~> 1.5",
      start_permanent: Mix.env == :prod,
      deps: deps(),
      name: "RFM69",
      source_url: "https://github.com/infinity-aps/elixir_rfm69",
      description: description(),
      package: package(),
      preferred_cli_env: [
        "dialyzer": :test
      ]
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp deps do
    [
      {:elixir_ale, "~> 1.0"},
      {:credo, "~> 0.8", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 0.5.1", only: :test, runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev}
    ]
  end

  defp description do
    """
    RFM69 is a library to handle sub-GHz wireless packet communication via an RFM69 chip.
    """
  end

  defp package do
    [
      maintainers: ["Timothy Mecklem"],
      licenses: ["MIT License"],
      links: %{"Github" => "https://github.com/infinity-aps/elixir_rfm69"}
    ]
  end
end


lib/rfm69.ex
defmodule RFM69 do
  use Bitwise

  alias RFM69.Device
  alias RFM69.Configuration

  @moduledoc false
  @fifo_size 66
  @transfer_sleep trunc(@fifo_size / 4)

  @doc """
  Sets the base frequency of the RFM69 chip in MHz
  """
  @frf_location 0x07
  @spec set_base_frequency(Device.t, pos_integer) :: {:ok}
  def set_base_frequency(%Device{name: name}, mhz) do
    frequency_in_hz = trunc(mhz * 1_000_000)
    register_values = Configuration.frequency_to_registers(frequency_in_hz)
    register_bytes = <<register_values::24>>
    :ok = GenServer.call(name, {:write_burst, @frf_location, register_bytes})
    {:ok}
  end

  @doc """
  reads a frame of binary data from the chip's FIFO queue
  """
  @spec read(Device.t, non_neg_integer) :: {:ok, %{data: binary, rssi: number}} | {:error, any()}
  def read(%Device{name: name}, timeout_ms) do
    set_auto_modes(name, [])
    set_mode(name, :receiver)
    GenServer.call(name, {:await_interrupt})

    result =
      receive do
        {:ok, :interrupt_received} ->
          rssi = read_rssi(name)
          bytes = read_until_null(name, <<>>)
          {:ok, %{data: bytes, rssi: rssi}}
      after
        timeout_ms ->
          GenServer.call(name, {:cancel_interrupt})
          {:error, :timeout}
      end

    set_mode(name, :sleep)
    result
  end

  @doc """
  Writes a frame of binary data to the FIFO queue
  """
  @spec write(Device.t, binary, pos_integer, non_neg_integer, non_neg_integer, boolean) :: {:ok, <<>>}
  def write(chip = %Device{name: name}, packet_bytes, repetitions, repetition_delay, timeout_ms, initial \\ true) do
    if initial == true do
      clear_fifo(name)
      set_mode(name, :standby)
    end

    case repetitions do
      r when r >= 1 ->
        _write(name, packet_bytes, timeout_ms)
        :timer.sleep(repetition_delay)
        write(chip, packet_bytes, repetitions - 1, repetition_delay, timeout_ms, false)

      0 ->
        set_mode(name, :sleep)
        {:ok, ""}
    end
  end

  @doc """
  Writes the given binary data to the FIFO queue and waits for the response.
  """
  @spec write_and_read(Device.t, binary, non_neg_integer) :: {:ok, %{data: binary, rssi: number}} | {:error, any()}
  def write_and_read(chip = %Device{}, packet_bytes, timeout_ms) do
    write(chip, packet_bytes, 1, 0, timeout_ms)
    read(chip, timeout_ms)
  end

  @spec clear_buffers(Device.t) :: :ok
  def clear_buffers(%Device{name: name}) do
    clear_fifo(name)
  end

  @doc """
  Writes a full configuration struct to the registers on the chip.
  """
  @configuration_start 0x01
  @spec write_configuration(Device.t, RFM69.Configuration.t) :: :ok
  def write_configuration(%Device{name: name}, rf_config = %Configuration{}) do
    configuration_bytes = Configuration.to_binary(rf_config)
    :ok = GenServer.call(name, {:write_burst, @configuration_start, configuration_bytes})
  end

  defp _write(name, packet_bytes, _timeout_ms) do
    modes = [:enter_condition_fifo_not_empty, :exit_condition_fifo_empty, :intermediate_mode_tx]
    set_auto_modes(name, modes)
    transmit(name, packet_bytes, @fifo_size)
  end

  defp read_until_null(name, data) do
    case GenServer.call(name, {:read_burst, 0x00, 1}) do
      {:ok, <<0x00::8>>} -> data
      {:ok, <<byte::8>>} -> read_until_null(name, data <> <<byte::8>>)
    end
  end

  defp transmit(name, bytes, available_buffer_bytes) when byte_size(bytes) <= available_buffer_bytes do
    _transmit(name, bytes)
    wait_for_mode(name, :standby)
  end

  defp transmit(name, bytes, available_buffer_bytes) do
    <<transmit_now::binary-size(available_buffer_bytes), transmit_later::binary>> = bytes
    available_buffer_bytes = _transmit(name, transmit_now)
    transmit(name, transmit_later, available_buffer_bytes)
  end

  defp _transmit(name, bytes) do
    GenServer.call(name, {:write_burst, 0x00, bytes})
    :timer.sleep(@transfer_sleep)
    wait_for_buffer_to_become_available(name)
  end

  @reg_irq_flags2 0x28
  @fifo_overrun 0x10
  defp clear_fifo(name) do
    GenServer.call(name, {:write_burst, @reg_irq_flags2, <<@fifo_overrun::8>>})
  end

  @fifo_level 0x20
  defp fifo_threshold_exceeded?(name) do
    {:ok, <<byte::8>>} = GenServer.call(name, {:read_burst, @reg_irq_flags2, 1})
    (byte &&& @fifo_level) != 0x00
  end

  defp wait_for_buffer_to_become_available(name) do
    case fifo_threshold_exceeded?(name) do
      true -> wait_for_buffer_to_become_available(name)
      false -> nil
    end
  end

  @reg_op_mode 0x01
  @modes %{
    sequencer_off: 0x80,
    listen_on:     0x40,
    listen_abort:  0x20,
    mode_shift:    0x02,
    mode_mask:     0x1C,
    sleep:         0x00,
    standby:       0x04,
    freq_synth:    0x08,
    transmitter:   0x0C,
    receiver:      0x10
  }
  defp set_mode(name, mode) do
    mode = Map.get(@modes, mode)
    current_mode = get_mode(name)

    case mode do
      ^current_mode -> true
      _ ->
        GenServer.call(name, {:write_burst, @reg_op_mode, <<mode::8>>})
        wait_for_mode_ready(name, mode)
    end
  end

  defp get_mode(name) do
    {:ok, <<byte>>} = GenServer.call(name, {:read_burst, @reg_op_mode, 1})
    byte
  end

  defp wait_for_mode(name, mode) do
    mode = Map.get(@modes, mode)
    _wait_for_mode(name, mode)
  end

  defp _wait_for_mode(name, mode) do
    if get_mode(name) != mode do
      :timer.sleep(1)
      _wait_for_mode(name, mode)
    end
  end

  @reg_auto_modes 0x3B
  @auto_modes %{
    enter_condition_fifo_not_empty: 0x20,
    exit_condition_fifo_empty:      0x04,
    intermediate_mode_tx:           0x03
  }
  defp set_auto_modes(name, modes) do
    register_value =
      modes
      |> Enum.map(fn mode -> @auto_modes[mode] end)
      |> Enum.reduce(0x00, &(&1 ||| &2))

    GenServer.call(name, {:write_burst, @reg_auto_modes, <<register_value::8>>})
  end

  @reg_irq_flags1 0x27
  @mode_ready     0x80
  defp wait_for_mode_ready(name, mode) do
    {:ok, <<current_mode>>} = GenServer.call(name, {:read_burst, @reg_op_mode, 1})
    {:ok, <<irq_flags1>>} = GenServer.call(name, {:read_burst, @reg_irq_flags1, 1})
    mode_ready = (irq_flags1 &&& @mode_ready) != 0x00

    case current_mode == mode && mode_ready do
      true -> true
      false -> wait_for_mode_ready(name, mode)
    end
  end

  @reg_rssi_val 0x24
  defp read_rssi(name) do
    {:ok, <<raw_rssi::8>>} = GenServer.call(name, {:read_burst, @reg_rssi_val, 1})
    -1 * trunc(raw_rssi) / 2
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :rfm69, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:rfm69, :key)
#
# You can also configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/rfm69/device.ex
defmodule RFM69.Device do
  @moduledoc """
  `RFM69.Device` is a GenServer to manage the SPI, reset and receive interrupt pids state and provide functions for reading and writing the registers and the FIFO buffer.
  """

  use Bitwise
  use GenServer

  require Logger

  alias ElixirALE.{GPIO, SPI}

  defstruct name: nil, device: nil, reset_pin: nil, interrupt_pin: nil
  @type t :: %RFM69.Device{name: :atom, device: String.t, reset_pin: integer, interrupt_pin: integer}

  @hardware_version 0x10

  @doc """
  Accepts the SPI device (eg "spidev0.0"), the reset and interrupt GPIO pin numbers and initializes communication with
  the RFM69 chip.
  """
  def start_link(%RFM69.Device{name: name, device: device, reset_pin: reset_pin, interrupt_pin: interrupt_pin}) do
    GenServer.start_link(__MODULE__, [device, reset_pin, interrupt_pin], name: name)
  end

  @doc """
  Initializes the SPI device and the reset and interrupt pins and prepares the RFM69 chip for interaction.
  """
  def init([spi_device, reset_pin, interrupt_pin]) do
    with {:ok, spi_pid} <- SPI.start_link(spi_device, speed_hz: 6_000_000),
         {:ok, reset_pid} <- GPIO.start_link(reset_pin, :output),
         {:ok, interrupt_pid} <- GPIO.start_link(interrupt_pin, :input),
         :ok <- GPIO.write(reset_pid, 0),
         {:ok, <<0x24>>} <- _read_burst(spi_pid, @hardware_version, 1) do

      {:ok, %{spi_pid: spi_pid, reset_pid: reset_pid, interrupt_pid: interrupt_pid}}
    else
      error ->
        message = "The SPI RFM69 chip interface failed to start: #{inspect error}"
        {:error, message}
    end
  end

  def chip_present?(%RFM69.Device{device: device}) do
    case SPI.start_link(device, speed_hz: 6_000_000) do
      {:ok, spi_pid} ->
        found = {:ok, <<0x24>>} == _read_burst(spi_pid, @hardware_version, 1)
        SPI.release(spi_pid)
        found
      _error -> false
    end
  end

  @doc """
  Sets the reset GPIO pin high long enough to power cycle the RFM69 chip
  """
  def handle_call({:reset}, _from, state = %{reset_pid: reset_pid}) do
    GPIO.write(reset_pid, 1)
    :timer.sleep(1)
    GPIO.write(reset_pid, 0)
    :timer.sleep(5)
    {:reply, :ok, state}
  end

  @doc """
  Writing to one or more configuration register locations is done via a "write burst", where the first byte transferred
  over SPI is the location logically ORed with 0x80 (the write bit), the second byte is the value for that register
  location, and each subsequent sent byte is stored in the proceeding register locations. The RFM69 chip auto-increments
  the register location for each byte so that a single frame of bytes can write the entire register configuration.
  """
  def handle_call({:write_burst, location, data}, _from, state = %{spi_pid: spi_pid}) do
    {:reply, _write_burst(spi_pid, location, data), state}
  end

  @doc """
  Read burst operations happen similarly to `write_burst`, where the first byte is the register location, and each
  subsequent byte sent (the value of the tx byte doesn't matter) results in a byte read with the value of that register
  location using the same auto-incrementing as in writes.
  """
  def handle_call({:read_burst, location, byte_count}, _from, state = %{spi_pid: spi_pid}) do
    {:reply, _read_burst(spi_pid, location, byte_count), state}
  end

  @doc """
  Sets the receive pin interrupt so that the next incoming packet triggers an `{:ok, :interrupt_received}` message to
  the caller so that the fifo buffer can be processed to decode the packet.
  """
  def handle_call({:await_interrupt}, {sender, _}, state = %{interrupt_pid: interrupt_pid}) do
    GPIO.set_int(interrupt_pid, :rising)
    {:reply, :ok, Map.put(state, :awaiting_interrupt, sender)}
  end

  @doc """
  Cancels the receive pin interrupt. The interrupt is canceled under normal conditions when a response message is
  triggered, but in the case of a timeout where the caller no longer has interest in the next packet, this function can
  be used.
  """
  def handle_call({:cancel_interrupt}, {_, _}, state = %{interrupt_pid: interrupt_pid}) do
    GPIO.set_int(interrupt_pid, :none)
    {:reply, :ok, Map.delete(state, :awaiting_interrupt)}
  end

  def handle_info(
        {:gpio_interrupt, _, :rising},
        state = %{awaiting_interrupt: awaiting_interrupt, interrupt_pid: interrupt_pid}
      ) do
    GPIO.set_int(interrupt_pid, :none)
    send(awaiting_interrupt, {:ok, :interrupt_received})
    {:noreply, Map.delete(state, :awaiting_interrupt)}
  end

  def handle_info({:gpio_interrupt, _, _}, state) do
    {:noreply, state}
  end

  @write_mode 0x80
  defp _write_burst(spi_pid, location, data) do
    tx_bytes = <<@write_mode ||| location::8>> <> data
    SPI.transfer(spi_pid, tx_bytes)
    :ok
  end

  defp _read_burst(spi_pid, location, byte_count) do
    size_bits = byte_count * 8
    tx_bytes = <<location::8, 0x00::size(size_bits)>>
    <<_::8, rx_bytes::binary>> = SPI.transfer(spi_pid, tx_bytes)
    {:ok, rx_bytes}
  end
end


lib/rfm69/rf_configuration.ex
defmodule RFM69.RFConfiguration do
  @moduledoc """
  RFConfiguration is a module and associated struct to model the registers in an RFM69 chip
  """

  use Bitwise

  alias RFM69.RFConfiguration

  # default settings from
  # http://www.hoperf.com/upload/rf/RFM69HW-V1.3.pdf

  defstruct [op_mode:        0x04,               # Operating modes of the transceiver
             data_modul:     0x00,               # Data operation mode and Modulation settings
             bitrate:        0x1A0B,             # Bit Rate setting
             fdev:           0x0052,             # Frequency Deviation setting
             frf:            0xE4C000,           # RF Carrier Frequency
             osc1:           0x41,               # RF Oscillators Settings
             afc_ctrl:       0x00,               # AFC control in low modulation index situations
             reserved0_c:    0x02,
             listen1:        0x92,               # Listen Mode settings
             listen2:        0xF5,               # Listen Mode Idle duration
             listen3:        0x20,               # Listen Mode Rx duration
             version:        0x24,               # PA selection and Output Power control
             pa_level:       0x9F,               # PA selection and Output Power control
             pa_ramp:        0x09,               # Control of the PA ramp time in FSK mode
             ocp:            0x1A,               # Over Current Protection control
             reserved14:     0x40,
             reserved15:     0xB0,
             reserved16:     0x7B,
             reserved17:     0x9B,
             lna:            0x88,               # LNA settings
             rx_bw:          0x55,               # Channel Filter BW Control
             afc_bw:         0x8B,               # Channel Filter BW control during the AFC routine
             ook_peak:       0x40,               # OOK demodulator selection and control in peak mode
             ook_avg:        0x80,               # Average threshold control of the OOK demodulator
             ook_fix:        0x06,               # Fixed threshold control of the OOK demodulator
             afc_fei:        0x10,               # AFC and FEI control and status
             afc:            0x00,               # Frequency correction of the AFC
             fei:            0x00,               # Calculated frequency error
             rssi_config:    0x02,               # RSSI-related settings
             rssi_value:     0xFF,               # RSSI value in dBm
             dio_mapping1:   0x00,               # Mapping of pins DIO0 to DIO3
             dio_mapping2:   0x07,               # Mapping of pins DIO4 and DIO5, ClkOut frequency
             irq_flags1:     0x80,               # Status register: PLL Lock state, Timeout, RSSI > Threshold...
             irq_flags2:     0x00,               # Status register: FIFO handling flags...
             rssi_thresh:    0xE4,               # RSSI Threshold control
             rx_timeout1:    0x00,               # Timeout duration between Rx request and RSSI detection
             rx_timeout2:    0x00,               # Timeout duration between RSSI detection and PayloadReady
             preamble:       0x0003,             # Preamble length
             sync_config:    0x98,               # Sync Word Recognition control
             sync_value:     0x0101010101010101, # Sync Word bytes, 1 through 8
             packet_config1: 0x10,               # Packet mode settings
             payload_length: 0x40,               # Payload length setting
             node_adrs:      0x00,               # Node address
             broadcast_adrs: 0x00,               # Broadcast address
             auto_modes:     0x00,               # Auto modes settings
             fifo_thresh:    0x8F,               # Fifo threshold, Tx start condition
             packet_config2: 0x02,               # Packet mode settings
             aes_key:        0x00,               # 16 bytes of the cypher key
             temp1:          0x01,               # Temperature Sensor control
             temp2:          0x00,               # Temperature readout Omit test
             test_lna:       0x1B,               # Sensitivity boost
             test_pa1:       0x55,               # High Power PA settings
             test_pa2:       0x70,               # High Power PA settings
             test_dagc:      0x30,               # Fading Margin Improvement
             test_afc:       0x00]               # AFC offset for low modulation index AFC

  def to_binary(rf_config = %RFConfiguration{}) do
    <<rf_config.op_mode::8,
      rf_config.data_modul::8,
      rf_config.bitrate::16,
      rf_config.fdev::16,
      rf_config.frf::24,
      rf_config.osc1::8,
      rf_config.afc_ctrl::8,
      rf_config.reserved0_c::8,
      rf_config.listen1::8,
      rf_config.listen2::8,
      rf_config.listen3::8,
      rf_config.version::8,
      rf_config.pa_level::8,
      rf_config.pa_ramp::8,
      rf_config.ocp::8,
      rf_config.reserved14::8,
      rf_config.reserved15::8,
      rf_config.reserved16::8,
      rf_config.reserved17::8,
      rf_config.lna::8,
      rf_config.rx_bw::8,
      rf_config.afc_bw::8,
      rf_config.ook_peak::8,
      rf_config.ook_avg::8,
      rf_config.ook_fix::8,
      rf_config.afc_fei::8,
      rf_config.afc::16,
      rf_config.fei::16,
      rf_config.rssi_config::8,
      rf_config.rssi_value::8,
      rf_config.dio_mapping1::8,
      rf_config.dio_mapping2::8,
      rf_config.irq_flags1::8,
      rf_config.irq_flags2::8,
      rf_config.rssi_thresh::8,
      rf_config.rx_timeout1::8,
      rf_config.rx_timeout2::8,
      rf_config.preamble::16,
      rf_config.sync_config::8,
      rf_config.sync_value::64,
      rf_config.packet_config1::8,
      rf_config.payload_length::8,
      rf_config.node_adrs::8,
      rf_config.broadcast_adrs::8,
      rf_config.auto_modes::8,
      rf_config.fifo_thresh::8,
      rf_config.packet_config2::8,
      rf_config.aes_key::128,
      rf_config.temp1::8,
      rf_config.temp2::8>>
  end

  @oscillator_frequency 32_000_000
  def frequency_to_registers(frequency_in_hz) do
    trunc(((frequency_in_hz <<< 19) + (@oscillator_frequency / 2)) / @oscillator_frequency)
  end

  def bitrate_to_registers(bitrate) do
    trunc((@oscillator_frequency + bitrate / 2) / bitrate)
  end
end


lib/rfm69/configuration.ex
defmodule RFM69.Configuration do
  @moduledoc """
  RFM69.Configuration is a module and associated struct to model register configuration in an RFM69 chip
  """

  use Bitwise

  alias RFM69.Configuration

  # default settings from
  # http://www.hoperf.com/upload/rf/RFM69HW-V1.3.pdf
  defstruct [op_mode:        0x04,               # Operating modes of the transceiver
             data_modul:     0x00,               # Data operation mode and Modulation settings
             bitrate:        0x1A0B,             # Bit Rate setting
             fdev:           0x0052,             # Frequency Deviation setting
             frf:            0xE4C000,           # RF Carrier Frequency
             osc1:           0x41,               # RF Oscillators Settings
             afc_ctrl:       0x00,               # AFC control in low modulation index situations
             reserved0_c:    0x02,
             listen1:        0x92,               # Listen Mode settings
             listen2:        0xF5,               # Listen Mode Idle duration
             listen3:        0x20,               # Listen Mode Rx duration
             version:        0x24,               # PA selection and Output Power control
             pa_level:       0x9F,               # PA selection and Output Power control
             pa_ramp:        0x09,               # Control of the PA ramp time in FSK mode
             ocp:            0x1A,               # Over Current Protection control
             reserved14:     0x40,
             reserved15:     0xB0,
             reserved16:     0x7B,
             reserved17:     0x9B,
             lna:            0x88,               # LNA settings
             rx_bw:          0x55,               # Channel Filter BW Control
             afc_bw:         0x8B,               # Channel Filter BW control during the AFC routine
             ook_peak:       0x40,               # OOK demodulator selection and control in peak mode
             ook_avg:        0x80,               # Average threshold control of the OOK demodulator
             ook_fix:        0x06,               # Fixed threshold control of the OOK demodulator
             afc_fei:        0x10,               # AFC and FEI control and status
             afc:            0x00,               # Frequency correction of the AFC
             fei:            0x00,               # Calculated frequency error
             rssi_config:    0x02,               # RSSI-related settings
             rssi_value:     0xFF,               # RSSI value in dBm
             dio_mapping1:   0x00,               # Mapping of pins DIO0 to DIO3
             dio_mapping2:   0x07,               # Mapping of pins DIO4 and DIO5, ClkOut frequency
             irq_flags1:     0x80,               # Status register: PLL Lock state, Timeout, RSSI > Threshold...
             irq_flags2:     0x00,               # Status register: FIFO handling flags...
             rssi_thresh:    0xE4,               # RSSI Threshold control
             rx_timeout1:    0x00,               # Timeout duration between Rx request and RSSI detection
             rx_timeout2:    0x00,               # Timeout duration between RSSI detection and PayloadReady
             preamble:       0x0003,             # Preamble length
             sync_config:    0x98,               # Sync Word Recognition control
             sync_value:     0x0101010101010101, # Sync Word bytes, 1 through 8
             packet_config1: 0x10,               # Packet mode settings
             payload_length: 0x40,               # Payload length setting
             node_adrs:      0x00,               # Node address
             broadcast_adrs: 0x00,               # Broadcast address
             auto_modes:     0x00,               # Auto modes settings
             fifo_thresh:    0x8F,               # Fifo threshold, Tx start condition
             packet_config2: 0x02,               # Packet mode settings
             aes_key:        0x00,               # 16 bytes of the cypher key
             temp1:          0x01,               # Temperature Sensor control
             temp2:          0x00,               # Temperature readout Omit test
             test_lna:       0x1B,               # Sensitivity boost
             test_pa1:       0x55,               # High Power PA settings
             test_pa2:       0x70,               # High Power PA settings
             test_dagc:      0x30,               # Fading Margin Improvement
             test_afc:       0x00]               # AFC offset for low modulation index AFC

  @type t :: %Configuration{
    op_mode:        byte,
    data_modul:     byte,
    bitrate:        0..0xFFFF,
    fdev:           0..0xFFFF,
    frf:            0..0xFFFFFF,
    osc1:           byte,
    afc_ctrl:       byte,
    reserved0_c:    byte,
    listen1:        byte,
    listen2:        byte,
    listen3:        byte,
    version:        byte,
    pa_level:       byte,
    pa_ramp:        byte,
    ocp:            byte,
    reserved14:     byte,
    reserved15:     byte,
    reserved16:     byte,
    reserved17:     byte,
    lna:            byte,
    rx_bw:          byte,
    afc_bw:         byte,
    ook_peak:       byte,
    ook_avg:        byte,
    ook_fix:        byte,
    afc_fei:        byte,
    afc:            byte,
    fei:            byte,
    rssi_config:    byte,
    rssi_value:     byte,
    dio_mapping1:   byte,
    dio_mapping2:   byte,
    irq_flags1:     byte,
    irq_flags2:     byte,
    rssi_thresh:    byte,
    rx_timeout1:    byte,
    rx_timeout2:    byte,
    preamble:       0..0xFFFF,
    sync_config:    byte,
    sync_value:     0..0xFFFFFFFFFFFFFFFF,
    packet_config1: byte,
    payload_length: byte,
    node_adrs:      byte,
    broadcast_adrs: byte,
    auto_modes:     byte,
    fifo_thresh:    byte,
    packet_config2: byte,
    aes_key:        byte,
    temp1:          byte,
    temp2:          byte,
    test_lna:       byte,
    test_pa1:       byte,
    test_pa2:       byte,
    test_dagc:      byte,
    test_afc:       byte,
  }

  @doc """
  Converts a `Configuration.t` struct into a binary that maps to the corresponding RFM69HCW register bytes
  """
  @spec to_binary(Configuration.t) :: binary
  def to_binary(c = %Configuration{}) do
    <<c.op_mode::8,
      c.data_modul::8,
      c.bitrate::16,
      c.fdev::16,
      c.frf::24,
      c.osc1::8,
      c.afc_ctrl::8,
      c.reserved0_c::8,
      c.listen1::8,
      c.listen2::8,
      c.listen3::8,
      c.version::8,
      c.pa_level::8,
      c.pa_ramp::8,
      c.ocp::8,
      c.reserved14::8,
      c.reserved15::8,
      c.reserved16::8,
      c.reserved17::8,
      c.lna::8,
      c.rx_bw::8,
      c.afc_bw::8,
      c.ook_peak::8,
      c.ook_avg::8,
      c.ook_fix::8,
      c.afc_fei::8,
      c.afc::16,
      c.fei::16,
      c.rssi_config::8,
      c.rssi_value::8,
      c.dio_mapping1::8,
      c.dio_mapping2::8,
      c.irq_flags1::8,
      c.irq_flags2::8,
      c.rssi_thresh::8,
      c.rx_timeout1::8,
      c.rx_timeout2::8,
      c.preamble::16,
      c.sync_config::8,
      c.sync_value::64,
      c.packet_config1::8,
      c.payload_length::8,
      c.node_adrs::8,
      c.broadcast_adrs::8,
      c.auto_modes::8,
      c.fifo_thresh::8,
      c.packet_config2::8,
      c.aes_key::128,
      c.temp1::8,
      c.temp2::8 >>
  end

  @oscillator_frequency 32_000_000
  @doc """
  Converts a frequency in hz into the three-byte register format
  """
  @spec frequency_to_registers(non_neg_integer) :: 0..0xFFFFFF
  def frequency_to_registers(frequency_in_hz) do
    trunc(((frequency_in_hz <<< 19) + @oscillator_frequency / 2) / @oscillator_frequency)
  end

  @doc """
  Converts bitrate into the two byte register format
  """
  @spec bitrate_to_registers(non_neg_integer) :: 0..0xFFFF
  def bitrate_to_registers(bitrate) do
    trunc((@oscillator_frequency + bitrate / 2) / bitrate)
  end
end
